p2p/p2p.go

```diff
@@ -51,134 +66,248 @@ func NewService(conf *Config, msgSend chan<- Message, msgRec <-chan Message) (*S
 	}
 
-	s.host.broadcast(encodedMsg)
-
-	return err
+	return true
 }
 
-// handleStream handles the stream, and rebroadcasts the message based on it's type
+// handleStream parses the message written to the data stream and calls the
+// associated message handler (status or non-status) based on message type
 func (s *Service) handleStream(stream net.Stream) {
-	msg, _, err := parseMessage(stream)
+
+	// parse message and exit on error
+	msg, err := parseMessage(stream)
 	if err != nil {
-		return
+		log.Debug("parse message", "err", err)
+		return // exit on error
 	}
 
-	// Write the message back to channel
+	log.Trace(
+		"handle stream",
+		"host", stream.Conn().LocalPeer(),
+		"peer", stream.Conn().RemotePeer(),
+		"type", msg.GetType(),
+	)
+
+	if msg.GetType() == StatusMsgType {
+		// handle status message type
+		s.handleStreamStatus(stream, msg)
+	} else {
+		// handle other message types
+		s.handleStreamNonStatus(stream, msg)
+	}
+
+	// send message to core service
 	s.msgSend <- msg
+}
+
+// handleStreamStatus handles status messages written to the stream
+func (s *Service) handleStreamStatus(stream network.Stream, msg Message) {
+
+	// TODO: use generated status message
+	hostStatus := statusMessage
+
+	switch {
+
+	// TODO: implement status message validation
+	case hostStatus.String() == msg.String():
+		log.Trace(
+			"status match",
+			"host", stream.Conn().LocalPeer(),
+			"peer", stream.Conn().RemotePeer(),
+		)
+
+		s.host.peerStatus[stream.Conn().RemotePeer()] = true
+
+	default:
+		log.Debug(
+			"status mismatch",
+			"host", stream.Conn().LocalPeer(),
+			"peer", stream.Conn().RemotePeer(),
+		)
 
-	// Rebroadcast all messages except for status messages
-	if msg.GetType() != StatusMsgType {
-		err = s.Broadcast(msg)
+		s.host.peerStatus[stream.Conn().RemotePeer()] = false
+
+		// close connection with peer if status message is not valid
+		err := s.host.h.Network().ClosePeer(stream.Conn().RemotePeer())
 		if err != nil {
-			log.Debug("failed to broadcast message: ", err)
-			return
+			log.Debug("close peer", "err", err)
 		}
+
 	}
 }
 
-var _ module.P2pApi = &Service{}
+// handleStreamNonStatus handles non-status messages written to the stream
+func (s *Service) handleStreamNonStatus(stream network.Stream, msg Message) {
+
+	status := s.host.peerStatus[stream.Conn().RemotePeer()]
+
+	// ignore message if peer status message has not been confirmed
+	if !status {
+		log.Trace(
+			"message ignored",
+			"host", stream.Conn().LocalPeer(),
+			"peer", stream.Conn().RemotePeer(),
+			"type", msg.GetType(),
+		)
+		return
+	}
+
+	// check if message should be broadcasted
+	if !s.shouldBroadcast(msg) {
+		log.Trace(
+			"message ignored",
+			"host", s.host.id(),
+			"type", msg.GetType(),
+		)
+		return
+	}
+
+	// broadcast to all connected peers if gossip enabled
+	if !s.host.noGossip {
 
-// ID returns the host's ID
+		log.Trace(
+			"gossiping",
+			"host", stream.Conn().LocalPeer(),
+			"type", msg.GetType(),
+		)
+
```

Comment:
Also seems unnecessary, once gossiping is migrated this might have more relevance

Reply:
Once again, it was helpful at the time when writing tests but I can remove.